conformance/tests/gateway-dynamic-listeners.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+	"sigs.k8s.io/gateway-api/conformance/utils/tls"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, GatewayListenerDynamicPorts)
+}
+
+var GatewayListenerDynamicPorts = suite.ConformanceTest{
+	ShortName: "GatewayListenerDynamicPorts",
+	Features: []suite.SupportedFeature{
+		suite.SupportGateway,
+		suite.SupportGatewayListenerDynamicPorts,
```

Comment:
Do we need "supported features" or similar for protocols? This specific test would fail with a L4 Gateway implementation which I don't think we want. Maybe there's a way to make the protocol(s) used in this test configurable to avoid this?

Reply:
> This specific test would fail with a L4 Gateway implementation which I don't think we want.

FWIW this doesn't seem isolated to this PR - there are a bunch of Gateway tests that are bundling HTTPRoutes in their fixtures.

> Do we need "supported features" or similar for protocols?  ... Maybe there's a way to make the protocol(s) used in this test configurable to avoid this?

I'm a bit mixed here - I think testing this across multiple protocols is good. I think making this test configurable isn't ideal since it's not exhaustively testing the different scenarios. Then it would require multiple invocations of the conformance tests to get a picture if this feature works for all protocols.

I think we want multiple tests and that for now I should add `SupportHTTPRoute` to the set of features here.

Reply:
Yeah agree with all of that, `SupportHTTPRoute` is probably the best short term solution.

Reply:
Added the `SupportHTTPRoute` feature to the `SupportedFeatures` of the test